core/src/main/java/io/grpc/internal/SharingExecutorService.java

```diff
@@ -0,0 +1,119 @@
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.AbstractExecutorService;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Wraps an existing executor, that keeps track of only the tasks it has added.
+ */
+public final class SharingExecutorService extends AbstractExecutorService {
+  private final Map<CleanupTask, Future<?>> activeTasks = new HashMap<CleanupTask, Future<?>>();
```

Comment:
it is, but it may not be the same future that was returned by the executor.  Any pre or post work from the delegate when wrapping an existing runnable would be lost.